libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -112,17 +113,48 @@ void AP_Mission::resume()
         return;
     }
 
-    // restart active navigation command. We run these on resume()
-    // regardless of whether the mission was stopped, as we may be
-    // re-entering AUTO mode and the nav_cmd callback needs to be run
-    // to setup the current target waypoint
+    // rewind the mission wp if the restart parameter is set negative
+    if (_restart < 0){
+        // ensure that reset parameter is not outside the limit of _wp_history_array
+        if (_restart < AP_MISSION_MAX_WP_HISTORY*-1){
+            gcs().send_text(MAV_SEVERITY_INFO, "Warning: MIS_RESTART below -%i limit",AP_MISSION_MAX_WP_HISTORY);
+            _restart = -1*AP_MISSION_MAX_WP_HISTORY;
+        }
+
+        // set resume flag to prevent hstory being re-written
+        _flags.resuming_mission = true;
+
+        // guard against _history_array_index < 0 to access _wp_history_array
+        if (_history_array_index < 0){
+            _history_array_index = 1;
+        }
+
+        // check if history buffer index needs to be wrapped
+        uint8_t resume_index;
+        if (_history_array_index < abs(_restart+1)){
```

Comment:
I'm not convinced the limits here are correct. Imagine _restart is -1000, then won't we go past the end of the history array?